test/util/PageRangeTest.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Xournal++
+ *
+ * This file is part of the Xournal UnitTests
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include <cppunit/extensions/HelperMacros.h>
+
+#include "PageRange.h"
+
+
+bool operator==(const PageRangeEntry& lhs, const PageRangeEntry& rhs) {
+    return lhs.first == rhs.first && lhs.last == rhs.last;
+}
+
+std::ostream& operator<<(std::ostream& os, const PageRangeEntry& val) {
+    return os << "PageRangeEntry{" << val.first << "," << val.last << "}";
+}
+
+class PageRangeTest: public CppUnit::TestFixture {
+    CPPUNIT_TEST_SUITE(PageRangeTest);
+
+    CPPUNIT_TEST(testParse);
+
+    CPPUNIT_TEST_SUITE_END();
+
+public:
+    void testParse() {
+        std::string ranges = "1-3, 6; 4-  : -5,, 6-5, 99-, -";
```

Comment:
You should add some more different ranges to this test. Try to think of corner cases. For example, what will happen if:

- One of the ranges is invalid but all others are valid
- No range at all is given
- Invalid characters are provided (letters, ...)
- ...

Reply:
What do you mean by invalid range? I'll try these.

Reply:
Should you still be confused, `--1;5-6` would be an example. Further examples are:

- `;2-`
- `-1-`

Reply:
`-1-` is not bad input, because everything after a valid read is ignored (that was the old behavior too).

It should however be documented, and of course the unit test should show it.

Reply:
OK I was right after all. I documented the behavior though in the function documentation string, and added those testcases as well.

Reply:
Please modify the behavior to not accept any range if there is an error parsing it. The reason is, that a user does not purposefully input an invalid range string. Hence, if you have an error while parsing it, it will most likely be a typo by the user. At such a time we should throw an error instead of silently continuing. Otherwise, users might end up with a PDF that has the wrong pages without knowing about it.

Reply:
Please also confirm that the GUI notifies the user in the event of such a wrong input.

Reply:
> Please modify the behavior to not accept any range if there is an error parsing it. 

When does this happen? I'm not sure what you refer to.

Reply:
I'm talking about the general behavior of `PageRange::parse`. It should not accept faulty input. Instead, it should warn the user that the input could not be parsed.

Reply:
Ah, OK I understand.

When is this function used? I don't know how it is used, when does the user get to interact with it?

Reply:
If I'm not mistaken this is part of the printing dialog. Your IDE might help you out here. Most modern IDEs allow you to search for functions that use another function. Just right-click on `PageRange::parse` and take a look at the context menu.

Reply:
I think the place that you want it to warn the user is in `src/gui/dialog/ExportDialog.cpp`, where it is called by `ExportDialog::getRange()`. Do you know which user action invokes this function?

Would the warning have to use gtk or does xournal have a function to show a warning message?

Reply:
1. Go to `File -> Export as...`
2. Select in the file dialog as file type `PNG`
3. Save the file

This will open the dialog you refer to. The best would be if you disable the `OK` button if the ranges are wrong and display a warning sign within the input field. You can use [gtk_entry_set_icon_from_pixbuf ()](https://developer-old.gnome.org/gtk3/stable/GtkEntry.html#gtk-entry-set-icon-from-pixbuf) for that.

 Another dialog on top seems to not be the best experience.

Reply:
For an appropriate icon you best speak with @luyatshimbalanga

Reply:
Or @luyatshimbalanga 

Both have designed icons for us previously.

Reply:
OK I've done some work on this. Try it out if you'd like and let me know what you think of it. I avoided using an icon, instead I used `gtk_style_context_*_class(context, "error");` to add/remove the error style in the text form. I also grayed out the OK button when appropriate. 

Since you probably last checked, I've also changed the parse function, now it emits exceptions in case of an issue, such as a pageCount equal to 0, or when the argument string is not a valid page range.

I still have a few minor things I'd like to change, i.e. I'd like to add a few more test cases, and some other minor things, but let me know what you think of it so far.

Reply:
Nice! Your work looks good from the first look at your code. Adding the context class is also a nice touch. I'll try it out live tomorrow.